### PR TITLE
Cell cycle scoring

### PR DESCRIPTION
**Pull request by ogibson**
# Description

# Details
#### URL to issue
N/A

#### Link to staging deployment URL (or set N/A)
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [x] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [x] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [x] Relevant Github READMEs updated **or** no GitHub README updates required.
- [x] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: r/R/cell_cycle_scoring.R [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+#' @param req
+#'
+#' @return
+#' @export
+#'
+#' @examples
+run_cell_cycle_scoring <- function(scdata, req) {
+  message("Running Cell Cycle Scoring")
+
+  s.gene.names <- Seurat::cc.genes$s.genes
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: variables use underscores!

File: r/R/cell_cycle_scoring.R [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+#' run_cell_cycle_scoring
+#'
+#' @param scdata
+#' @param req
+#'
+#' @return
+#' @export
+#'
+#' @examples
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: add docs, remove @examples line to avoid roxygen warnings.

also run devtools::document() to generate the manpages for the functions

File: r/R/cell_cycle_scoring.R [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ensure_is_list_in_json <- function(vector) {
+  if (length(vector) <= 1) {
+    return(as.list(vector))
+  } else {
+    return(vector)
+  }
+}
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: this function should be implemented somewhere already. Isn't it?

Comment (gerbeldo) [reply r/R/cell_cycle_scoring.R]: also, always remove the '@example' line, to avoid roxygen2 complaints

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  expect_true(result$key == "Phase")
+  expect_true(result$name == "Phase")
+  expect_true(result$type == "cellSets")
+
+  expect_equal(length(result$children), 1)
+  expect_true(result$children[[1]]$key == "Phase-S")
+  expect_true(result$children[[1]]$name == "S")
+})
+
+# Start writing tests
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]: remove gpt instructions 👀

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+          {
+            scdata <- mock_scdata()
+
+            result <- run_cell_cycle_scoring(scdata, req)
+
+            expect_is(result, "data.frame")
+
+            expect_true("cluster" %in% colnames(result))
+            expect_true("cell_ids" %in% colnames(result))
+          })
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]: fix indenting

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            # Check the structure of each child
+            for (child in result$children) {
+              expect_true("key" %in% names(child))
+              expect_true("name" %in% names(child))
+              expect_true("rootNode" %in% names(child))
+              expect_true("type" %in% names(child))
+              expect_true("color" %in% names(child))
+              expect_true("cellIds" %in% names(child))
+            }
+          })
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]: fix indents

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  )
+
+  pbmc_small <- SeuratObject::CreateSeuratObject(counts = counts)
+
+  if (phase == "S") {
+    pbmc_small@assays$RNA@counts[1:40, ] <- 3
+    pbmc_small@assays$RNA@counts[41:121, ] <- 0
+  } else{
+    pbmc_small@assays$RNA@counts[41:121, ] <- 3
+  }
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]: this is a bit weird. I understand that you're trying to over express phase-specific genes? in that case, and since you have rownames, you could re-write as  

```suggestion
  if (phase == "S") {
    pbmc_small@assays$RNA@counts[Seurat::cc.genes$s.genes, ] <- 3
    pbmc_small@assays$RNA@counts[Seurat::cc.genes$g2m.genes, ] <- 0
  } else {
    pbmc_small@assays$RNA@counts[Seurat::cc.genes$g2m.genes, ] <- 3
  }
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+  gene_annotations <- data.frame(
+    input = enids,
+    name = new_names,
+    original_name = row.names(counts),
+    row.names = enids
+  )
+
+  pbmc_small <- SeuratObject::CreateSeuratObject(counts = counts)
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]: this is not pbmc_small, but a randomly generated dataset, I would change the var name. Avoid confusion, since Seurat provides a pbmc_small dataset and we use it throughout the tests.

File: r/tests/testthat/test-cell_cycle_scoring.R (hunk 7 above)
Comment (gerbeldo) [reply r/tests/testthat/test-cell_cycle_scoring.R]: you could add a comment as well.

File: r/R/cell_cycle_scoring.R [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      type = "cellSets",
+      color = color_pool[1],
+      cellIds = ensure_is_list_in_json(unname(cells))
+    )
+    color_pool <- color_pool[-1]
+    cell_sets_object$children <-
+      append(cell_sets_object$children, list(new_set))
+  }
+  return(cell_sets_object)
+}
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: why is this here? isn't it in the clustering file?

Comment (gerbeldo) [reply r/R/cell_cycle_scoring.R]: in any case, the name sounds kind of wrong. maybe `format_cellcycle_cellsets`

File: r/R/cell_cycle_scoring.R [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+#' @param clustering_method
+#' @param color_pool
+#' @param name
+#'
+#' @return Formats cellsets dataframe into platform compatible object
+#' @export
+#'
+#' @examples
+format_cluster_cellsets <- function(cell_sets,
+                                    clustering_method,
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: this is confusing. you're passing 'Phase' here

File: r/R/cell_cycle_scoring.R [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                                    color_pool,
+                                    name = paste0(clustering_method)) {
+  message("Formatting cluster cellsets.")
+
+  # careful with capital l on type for the key.
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: this comment does not make sense here

File: r/R/cell_cycle_scoring.R [hunk 12]
```diff
@@ -0,0 +1,132 @@
+#' CellCyCleScoring
+#'
+#' @param req
+#' @param data
+#'
+#' @return Formatted cellsets
+#' @export
+#'
+#' @examples
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: delete examples line

File: r/R/cell_cycle_scoring.R [hunk 13]
```diff
@@ -0,0 +1,132 @@
+#' CellCyCleScoring
+#'
+#' @param req
+#' @param data
+#'
+#' @return Formatted cellsets
+#' @export
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: you should minimally fill things here. add a description line/paragraph, what are req and data, etc. this goes for all the PR

File: r/R/cell_cycle_scoring.R [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+  phase <- tryCatch({
+    Seurat::CellCycleScoring(
+      scdata,
+      s.features = s.genes,
+      g2m.features = g2m.genes,
+      set.ident = TRUE
+    )@meta.data$Phase
+  }, error = function(err) {
+    rep("Undetermined", ncol(scdata))
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: would this error message be clear enough?

**Review by gerbeldo — CHANGES_REQUESTED** [review]

File: r/R/cell_cycle_scoring.R (hunk 14 above)
Comment (ogibson) [reply r/R/cell_cycle_scoring.R]: It's not an error message though, it will just create a single "undetermined" cellset.

Comment (ogibson) [reply r/R/cell_cycle_scoring.R]: Maybe we should use a different word here. What do you think?

File: r/tests/testthat/test-cell_cycle_scoring.R (hunk 7 above)
Comment (ogibson) [reply r/tests/testthat/test-cell_cycle_scoring.R]: The counts don't have the symbols used in cc.genes, I will clarify with a comment.

File: r/tests/testthat/test-cell_cycle_scoring.R (hunk 4 above)
Comment (ogibson) [reply r/tests/testthat/test-cell_cycle_scoring.R]: It's the only one that slipped damn

File: r/R/cell_cycle_scoring.R [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+#'
+#' @param cell_sets Data frame with cell id and Phase
+#' @param color_pool
+#' @param name
+#'
+#' @return Formats cellsets dataframe into platform compatible object
+#' @export
+#'
+format_phase_cellsets <- function(cell_sets,
+                                    color_pool) {
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
format_phase_cellsets <- function(cell_sets,
                                  color_pool) {
```

File: r/R/cell_cycle_scoring.R (hunk 14 above)
Comment (gerbeldo) [reply r/R/cell_cycle_scoring.R]: maybe something more explicit? cc_phase_undetermined? or something like that. We will not be throwing an error, and a new cellset will be created. That cellset should have a clear name, so that when you see it in data exploration you know where it comes from.

File: r/R/cell_cycle_scoring.R [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+      s.features = s_genes,
+      g2m.features = g2m_genes,
+      set.ident = TRUE
+    )@meta.data$Phase
+  }, error = function(err) {
+    rep("Undetermined", ncol(scdata))
+  })
+
+  cellSets <-
+    data.frame(cluster = phase,
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: I would rename this to phase, since we're annotating cells, and not really clustering. I understand where this comed from though. To make it easier, I left change suggestions all over the PR.

File: r/R/cell_cycle_scoring.R [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+  cell_sets_object <-
+    list(
+      key = "Phase",
+      name = "Phase",
+      rootNode = TRUE,
+      type = "cellSets",
+      children = list()
+    )
+  for (cluster in unique(cell_sets$cluster)) {
+    cells <- cell_sets[cell_sets$cluster == cluster, "cell_ids"]
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: and phase here as well

File: r/R/cell_cycle_scoring.R [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+#' @param cell_sets Data frame with cell id and Phase
+#' @param color_pool
+#' @param name
+#'
+#' @return Formats cellsets dataframe into platform compatible object
+#' @export
+#'
+format_phase_cellsets <- function(cell_sets,
+                                    color_pool) {
+  message("Formatting cluster cellsets.")
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]: change message

Comment (gerbeldo) [reply r/R/cell_cycle_scoring.R]:
```suggestion
  message("Formatting cell cycle phase cellsets.")
```

File: r/R/cell_cycle_scoring.R (hunk 17 above)
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
  for (phase in unique(cell_sets$phase)) {
    cells <- cell_sets[cell_sets$phase == phase, "cell_ids"]
```

File: r/R/cell_cycle_scoring.R [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+      rootNode = TRUE,
+      type = "cellSets",
+      children = list()
+    )
+  for (cluster in unique(cell_sets$cluster)) {
+    cells <- cell_sets[cell_sets$cluster == cluster, "cell_ids"]
+
+    new_set <- list(
+      key = paste0("Phase", "-", cluster),
+      name = cluster,
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
      key = paste0("Phase", "-", phase),
      name = phase,
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+})
+
+test_that("run_cell_cycle_scoring returns a data frame with correct columns", {
+  scdata <- mock_scdata()
+
+  result <- suppressWarnings(run_cell_cycle_scoring(scdata))
+
+  expect_s3_class(result, "data.frame")
+
+  expect_true("cluster" %in% colnames(result))
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
  expect_true("phase" %in% colnames(result))
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+})
+
+test_that("run_cell_cycle_scoring properly classifies S cells", {
+  scdata <- mock_scdata("S")
+
+  result <- suppressWarnings(run_cell_cycle_scoring(scdata))
+
+  expect_s3_class(result, "data.frame")
+
+  expect_true(all(result$cluster == "S"))
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
  expect_true(all(result$phase == "S"))
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+})
+
+test_that("run_cell_cycle_scoring properly classifies G2M cells", {
+  scdata <- mock_scdata("G2M")
+
+  result <- suppressWarnings(run_cell_cycle_scoring(scdata))
+
+  expect_s3_class(result, "data.frame")
+
+  expect_true(all(result$cluster == "G2M"))
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
  expect_true(all(result$phase == "G2M"))
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+})
+
+test_that("run_cell_cycle_scoring returns 'Undetermined' cellset when no genes are detected.", {
+  scdata <- mock_scdata(phase = "S", add_cycle_genes = FALSE)
+
+  result <- suppressWarnings(run_cell_cycle_scoring(scdata))
+
+  expect_s3_class(result, "data.frame")
+
+  expect_true(all(result$cluster == "Undetermined"))
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
  expect_true(all(result$phase == "Undetermined"))
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+})
+
+test_that("format_phase_cellsets returns a list with the correct structure", {
+  # Create a mock data frame for testing
+  mock_cell_sets <- data.frame(
+    cluster = c("A", "A", "B", "C", "C"),
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
    phase = c("A", "A", "B", "C", "C"),
```

File: r/tests/testthat/test-cell_cycle_scoring.R [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+    expect_true("rootNode" %in% names(child))
+    expect_true("type" %in% names(child))
+    expect_true("color" %in% names(child))
+    expect_true("cellIds" %in% names(child))
+  }
+})
+
+test_that("format_phase_cellsets returns expected formatting", {
+  mock_cell_sets <- data.frame(
+    cluster = c("A", "A", "B", "C", "C"),
```
Comment (gerbeldo) [inline r/tests/testthat/test-cell_cycle_scoring.R]:
```suggestion
    phase = c("A", "A", "B", "C", "C"),
```

File: r/R/cell_cycle_scoring.R [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+#' run_cell_cycle_scoring
+#'
+#' Uses the Seurat method to calculate cell cycle scores.
+#'
+#' @param scdata SeuratObject
+#'
+#' @return Dataframe with cell ids and cycle stage prediction, or undetermined if there is an error.
+#' @export
+#'
+#' @examples
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
```

File: r/R/cell_cycle_scoring.R [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+  return(formatted_cell_sets)
+}
+
+#' run_cell_cycle_scoring
+#'
+#' Uses the Seurat method to calculate cell cycle scores.
+#'
+#' @param scdata SeuratObject
+#'
+#' @return Dataframe with cell ids and cycle stage prediction, or undetermined if there is an error.
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
#' @return data.frame with cell ids and cycle stage prediction, or undetermined if there is an error.
```

File: r/R/cell_cycle_scoring.R [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+
+#' format_phase_cellsets
+#'
+#' @param cell_sets Data frame with cell id and Phase
+#' @param color_pool
+#' @param name
+#'
+#' @return Formats cellsets dataframe into platform compatible object
+#' @export
+#'
```
Comment (gerbeldo) [inline r/R/cell_cycle_scoring.R]:
```suggestion
#' Format cell cycle phase cellsets
#'
#' @param cell_sets data.frame with cell ids and phase
#' @param color_pool list
#' @param name cell cycle phase name
#'
#' @return list of structured cellsets
#' @export
#'
```

**Review by gerbeldo — APPROVED** [review]
